www/blog/2023-01-08-tinyrpc-client.mdx

```diff
@@ -0,0 +1,210 @@
+  TProcedure extends AnyQueryProcedure
+    ? {
+        query: Resolver<TProcedure>;
+      }
+    : TProcedure extends AnyMutationProcedure
+    ? {
+        mutate: Resolver<TProcedure>;
+      }
+    : never;
+```
+
+Finally, we'll create a type that will recursively traverse the router and decorate the procedures with the `.query` and `.mutate` methods.
+
+```ts
+type DecoratedProcedureRecord<TProcedures extends ProcedureRouterRecord> = {
+  [TKey in keyof TProcedures]: TProcedures[TKey] extends AnyRouter
+    ? DecoratedProcedureRecord<TProcedures[TKey]['_def']['record']>
+    : TProcedures[TKey] extends AnyProcedure
+    ? DecorateProcedure<TProcedures[TKey]>
+    : never;
+};
+```
+
+Let's digest this type a bit:
+
+1. We pass a `ProcedureRouterRecord` to the type as a generic, which is a type containing all the procedures and sub-routers that exists on a tRPC router.
+2. We iterate over the keys of the record, which are the procedure or router names, and do the following:
+   1. If the key maps to a router, we recursively call the type on that router's procedure record, which will decorate all the procedures in that router. This will provide autocompletion as we traverse the path.
+   2. If the key maps to a procedure, we decorate the procedure using the `DecorateProcedure` type we created earlier.
+   3. If the key doesn't map to a procedure or router, we assign the `never` type which is like saying "this key doesn't exist" which will cause a type error if we try to access it.
+
+Finally, we'll need to use these types as we create the client. We'll add a generic to the `createTinyRPCClient` function that will accept any tRPC router type, and then we'll cast the return type of the function to the `DecoratedProcedureRecord` type we created earlier:
+
+```ts
+export const createTinyRPCClient = <TRouter extends AnyRouter>(url: string) => {
+  return createRecursiveProxy(async (opts) => {
+    // ...
+  }, []) as DecoratedProcedureRecord<TRouter['_def']['record']>;
+};
+```
+
+And that's it! Create the client and provide it your server's url and you'll get full autocompletion and type safety when you call your procedures!
+
+```ts
+import { type AppRouter } from '~/server/router';
+
+const client = createTinyRPCClient<AppRouter>('http://localhost:3000/api/trpc');
+
+const post = await client.post.byId.query('123');
+```
+
+The full code for the client can be found [here](https://github.com/trpc/trpc/blob/f08e36476fc0f09d512f63bfc1b225b417058dd0/packages/tests/server/minitrpc.ts):
+
+## Conclusion
+
+I hope you enjoyed this article and learned something about how tRPC works. Obviously, there's a lot of stuff missing compared to the full client shipped in `@trpc/client` such as batching, subscriptions and links to alter the request flow, but if you want a really small client with just the bare minimum, this is all you need! We also didn't cover much of the server side of things today, maybe we'll cover that in a future article.
```

Comment:
Maybe make a list of reasons why you should **not use** this in favour of the official client.

> You should probably not use this to in favour of the `@trpc/client` which is only a couple of kbs bigger - it comes with a lot more flexibility than what we're showcasing here:
> 
> - Links
> - Procedure batching
> - WebSockets / subscriptions
> - Nice error handling
> - Data transformers
> - Edge cases handling like when we don't get a tRPC-compliant response